examples/antiDdos/antiDDoS.go

```diff
@@ -0,0 +1,176 @@
+// Copyright 2017 Intel Corporation.
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+// Below is an example of simple anti DDoS function. Type of atack is L4 flood.
+// DDoS attack is considered when only one packet is sent from each of multiple
+// source addresses with high frequrncy.
+// Packets are split on flows (NOT YANFF flows) – abstraction of all packets
+// came from one scr addr and src port.
+// Layer 4 anti DDoS filter will be based on 2 indicators:
+// * Rate of inter-arrival times within [0, 200 us) (IAT):
+// Number of packets came in [0, 200 us)  / number of packets came in
+// * Rate of flows having only one packet (PPF):
+// Number of flows having 1 packet / number of flows
+// If IAT > IATthreshold , PPF > PPFthreshold we consider DDOS is going.
+// If DDOS is going we drop incoming packets that were not seen earlier, but record them for future.
+
+package main
+
+import (
+	"flag"
+	"sync/atomic"
+	"time"
+
+	"github.com/intel-go/yanff/common"
+	"github.com/intel-go/yanff/flow"
+	"github.com/intel-go/yanff/packet"
+)
+
+const (
+	monitoringWindow       = 6                       // 6s - expiration time of flow in a table
+	size2InPow             = 29                      // 2**29 will be size of a table
+	size                   = uint32(1 << size2InPow) // table size
+	iatThreshold           = float64(0.9)
+	ppfThreshold           = float64(0.8)
+	p                      = uint32(16777619)   // used for hash calculation
+	hashInit               = uint32(2166136261) // used for hash calculation
+	timeStampMask          = ((1 << 14) - 1)    // 14 first bits mask for timestamp
+	getNumOfPacketsMask    = (3 << 14)          // 2 last bits mask for num of packets
+	setNumOfPacketsTo1Mask = (1 << 14)          // first 2 bits set to 01 (means 1 pkt)
+	setNumOfPacketsTo2Mask = (2 << 14)          // first 2 bits set to 10 (means 1+ pkt)
+	timeShiftToGetSeconds  = 30                 // pow(2, 30) is approximately 1e9
+	interval200usToNanos   = 200000             // 200 us converted to nanoseconds
+)
+
+var (
+	inport  int
+	outport int
+	// flowTable stores number of packets in flow, it takes
+	// 2 first bits from uint16,
+	// time of last packet is stored in seconds and
+	// takes 14 last bits of uint16
+	flowTable              [size]uint16
+	numOfFlows             = int32(1)
+	numOfOnePktFlows       = int32(0)
+	packetsInInterval200us = int64(0)
+	allPackets             = int64(1)
+	lastTime               = time.Now().UnixNano()
+	isDdos                 uint32
+)
+
+func flowHash(srcAddr []byte, srcAddrLen int, srcPort uint32) uint32 {
+	hash := hashInit
+	for i := 0; i < srcAddrLen; i++ {
+		hash = (hash ^ uint32(srcAddr[i])) * p
+	}
+	hash += srcPort
+	hash += hash << 13
+	hash ^= hash >> 7
+	hash += hash << 3
+	hash ^= hash >> 17
+	hash += hash << 5
+	return hash & (size - 1)
+}
+
```

Comment:
Just curious, where did this hash function come from? Is it possible to use some more complicated hashing here (md5, sha, whatever) instead of this algorithm?

Reply:
It was https://ru.wikipedia.org/wiki/FNV and thanks for comment, there is the same "hash/fnv" in go!

Reply:
well...I can't explain why, but when I switch mine function to golang fnv, filter starts to drop 10% of good packets instead of 4... So I'll make some experiments